python/lsst/rubintv/production/watchers.py

```diff
@@ -0,0 +1,226 @@
+
+        Get the most recent expRecord for all the dataset types. These are
+        written to files for the FileWatchers to pick up.
+
+        Returns
+        -------
+        expRecords : `dict` [`str`, `lsst.daf.butler.DimensionRecord`]
+            A dict of the most recent exposure records, keyed by dataProduct.
+        """
+        expRecordDict = {}
+
+        # this where clause is no longer necessary for speed now that we are
+        # using order_by and limit, but it now functions to restrict data to
+        # being inside the range where exposures could ever realistically be
+        # taken, as some simulated data is "taken" in the year 3000. We must
+        # also restrict the exposure.id specifically, as the day_obs seems to
+        # be set to this millenium, but the expIds are like 3YYYMMDDNNNNN.
+        where = ('exposure.day_obs>20200101 and exposure.day_obs<21000101'
+                 ' and exposure.id<2100010100001')
+
+        for product in self.dataProducts:
+            # NB if you list multiple products for datasets= then it will only
+            # give expRecords for which all those products exist, so these must
+            # be done as separate queries
+            records = self.butler.registry.queryDimensionRecords("exposure",
+                                                                 datasets=product,
+                                                                 where=where)
+            records.order_by('-exposure.id')  # the minus means descending ordering
+            records.limit(1)
+            expRecordDict[product] = list(records)[0]
+        return expRecordDict
+
+    def run(self):
+        lastWrittenIds = {product: 0 for product in self.dataProducts}
+        while True:
+            try:
+                # get the new records for all dataproducts
+                newRecords = self._getLatestExpRecords()
+                # work out which ones are actually new and only write those out
+                found = {product: expRecord for product, expRecord in newRecords.items()
+                         if expRecord.id != lastWrittenIds[product]}
+
+                if not found:  # only sleep when there's nothing new at all
+                    sleep(self.cadence)
+                    # TODO: re-enable?
+                    # self.heartbeater.beat()
+                    continue
+                else:
+                    for product, expRecord in found.items():
+                        writeDataIdFile(self.locationConfig.dataIdScanPath, product, expRecord, log=self.log)
+                        lastWrittenIds[product] = expRecord.id
+                    # beat after the callback so as not to delay processing
+                    # and also so it is only called if things are working
+                    # TODO: re-enable?
+                    # self.heartbeater.beat()
+
+            except Exception as e:
+                raiseIf(self.doRaise, e, self.log)
```

Comment:
This is a big exception block, and it looks like there might be ways in which the `doRaise=False` branch could lead to this really hammering the butler hard if something goes wrong, which might lead to new errors, and so on.  Maybe add a `sleep` there, or even inside `raiseIf` to make sure that happens anywhere it might be needed?

And I think a common pattern is to increase that sleep duration as the failures continue - I certainly see GMail doing that to me, though this is well outside my area of expertise.

Reply:
Ah yes, good shout. And you're quite right, that pattern is, I believe, referred to as "exponential backoff" and is in fact implemented automatically in how the uploads are reattempted in failure in GCS (and in the `Uploader` class in this package). I don't think it's worth the hassle of coding it up here, but I've certainly added a sleep and code comment as to why it's there.